Test/git-issues/git-issue-2429.dfy

```diff
@@ -0,0 +1,16 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+predicate P<A>(s: seq<A>)
```

Comment:
Please also add tests for the other clauses. For example:

``` dafny
predicate P<A>(s: seq<A>)

trait T<A> {
  method M(a: A)
    requires P([a][0 := a])
    modifies if P([a][0 := a]) then {} else {this}
    ensures P([a][0 := a])
    decreases if P([a][0 := a]) then 3 else 4
  function F(a: A): int
    requires P([a][0 := a])
    reads if P([a][0 := a]) then {} else {this}
    ensures F(a) == 5 ==> P([a][0 := a])
    decreases if P([a][0 := a]) then 3 else 4
}

predicate Q<A>(s: seq<A>) ensures Q(s) ==> P(s)

class C extends T<object> {
  // A missing type substitution in the translator used to cause issues here:
  method M(a: object)
    requires Q([a][0 := a])
    modifies if Q([a][0 := a]) then {} else {this}
    ensures Q([a][0 := a])
    decreases if Q([a][0 := a]) then 3 else 2
  function F(a: object): int
    requires Q([a][0 := a])
    reads if Q([a][0 := a]) then {} else {this}
    ensures F(a) == 5 ==> Q([a][0 := a])
    decreases if Q([a][0 := a]) then 3 else 4
}
```

Curiously, to trigger the bug for `requiers`/`modifies`/`reads`/`ensures`, it was necessary only to include the specification clause in the `trait`. But for `decreases`, one had to include it in both the `trait` and the `class` in order for the malformed Boogie to appear.